File: packages/bridge/src/git/diff_parser.ts
import type { DiffFile, DiffHunk, DiffLine } from "../types";

const FILE_HEADER_RE = /^diff --git a\/(.*) b\/(.*)$/;
const OLD_FILE_RE = /^--- (?:a\/(.*)|\/(dev\/null))$/;
const NEW_FILE_RE = /^\+\+\+ (?:b\/(.*)|\/(dev\/null))$/;
const HUNK_HEADER_RE = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(.*)$/;

function determineFileStatus(
  isNew: boolean,
  isDeleted: boolean,
  isRenamed: boolean,
): DiffFile["status"] {
  if (isNew) {
    return "added";
  }
  if (isDeleted) {
    return "deleted";
  }
  if (isRenamed) {
    return "renamed";
  }
  return "modified";
}

export function parseDiff(rawDiff: string): DiffFile[] {
  const files: DiffFile[] = [];
  if (!rawDiff || rawDiff.trim().length === 0) {
    return files;
  }

  const lines = rawDiff.split("\n");
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    const fileHeaderMatch = FILE_HEADER_RE.exec(line);
    if (!fileHeaderMatch) {
      i++;
      continue;
    }

    const oldPathRaw = fileHeaderMatch[1];
    const newPathRaw = fileHeaderMatch[2];
    i++;

    let oldPath = oldPathRaw;
    let newPath = newPathRaw;
    let isNew = false;
    let isDeleted = false;
    let isRenamed = oldPath !== newPath;

    while (i < lines.length && lines[i].startsWith("index ")) {
      i++;
    }

    if (i < lines.length) {
      const oldMatch = OLD_FILE_RE.exec(lines[i]);
      if (oldMatch) {
        if (lines[i] === "--- /dev/null") {
          isNew = true;
        } else if (oldMatch[1]) {
          oldPath = oldMatch[1];
        }
        i++;
      }
    }

    if (i < lines.length) {
      const newMatch = NEW_FILE_RE.exec(lines[i]);
      if (newMatch) {
        if (lines[i] === "+++ /dev/null") {
          isDeleted = true;
        } else if (newMatch[1]) {
          newPath = newMatch[1];
        }
        i++;
      }
    }

    const hunks: DiffHunk[] = [];
    let additions = 0;
    let deletions = 0;

    while (i < lines.length) {
      const hunkLine = lines[i];

      if (FILE_HEADER_RE.test(hunkLine)) {
        break;
      }

      const hunkMatch = HUNK_HEADER_RE.exec(hunkLine);
      if (!hunkMatch) {
        i++;
        continue;
      }

      const oldStart = parseInt(hunkMatch[1], 10);
      const oldLines = hunkMatch[2] !== undefined ? parseInt(hunkMatch[2], 10) : 1;
      const newStart = parseInt(hunkMatch[3], 10);
      const newLines = hunkMatch[4] !== undefined ? parseInt(hunkMatch[4], 10) : 1;
      const header = hunkLine;
      i++;

      const diffLines: DiffLine[] = [];
      let oldLineNum = oldStart;
      let newLineNum = newStart;

      while (i < lines.length) {
        const diffLine = lines[i];

        if (FILE_HEADER_RE.test(diffLine) || HUNK_HEADER_RE.test(diffLine)) {
          break;
        }

        if (diffLine.startsWith("+")) {
          additions++;
          diffLines.push({
            type: "added",
            content: diffLine.slice(1),
            new_line_number: newLineNum++,
          });
        } else if (diffLine.startsWith("-")) {
          deletions++;
          diffLines.push({
            type: "removed",
            content: diffLine.slice(1),
            old_line_number: oldLineNum++,
          });
        } else if (diffLine.startsWith(" ") || diffLine === "") {
          diffLines.push({
            type: "context",
            content: diffLine.startsWith(" ") ? diffLine.slice(1) : diffLine,
            old_line_number: oldLineNum++,
            new_line_number: newLineNum++,
          });
        }

        i++;
      }

      hunks.push({
        old_start: oldStart,
        old_lines: oldLines,
        new_start: newStart,
        new_lines: newLines,
        header,
        lines: diffLines,
      });
    }

    files.push({
      path: isDeleted ? oldPath : newPath,
      old_path: oldPath,
      new_path: newPath,
      status: determineFileStatus(isNew, isDeleted, isRenamed),
      additions,
      deletions,
      hunks,
    });
  }

  return files;
}

File: apps/mobile/lib/shared/utils/diff_parser.dart
import '../../core/models/git_models.dart';

/// Parses unified diff strings into [DiffFile] objects.
class DiffParser {
  static final _fileHeaderOld = RegExp(r'^--- a/(.+)$');
  static final _fileHeaderNew = RegExp(r'^\+\+\+ b/(.+)$');
  static final _hunkHeader = RegExp(
    r'^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(.*)$',
  );

  /// Parse a unified diff string into a list of [DiffFile] objects.
  static List<DiffFile> parse(String rawDiff) {
    final files = <DiffFile>[];
    final lines = rawDiff.split('\n');

    String? oldPath;
    String? newPath;
    final hunks = <DiffHunk>[];

    List<DiffLine>? currentHunkLines;
    DiffHunk? currentHunk;
    int oldLineNum = 0;
    int newLineNum = 0;

    void flushHunk() {
      if (currentHunk != null && currentHunkLines != null) {
        hunks.add(DiffHunk(
          header: currentHunk!.header,
          oldStart: currentHunk!.oldStart,
          oldLines: currentHunk!.oldLines,
          newStart: currentHunk!.newStart,
          newLines: currentHunk!.newLines,
          lines: List.unmodifiable(currentHunkLines!),
        ));
        currentHunk = null;
        currentHunkLines = null;
      }
    }

    void flushFile() {
      if (oldPath != null && newPath != null) {
        flushHunk();

        int additions = 0;
        int deletions = 0;
        for (final hunk in hunks) {
          for (final line in hunk.lines) {
            if (line.type == DiffLineType.added) additions++;
            if (line.type == DiffLineType.removed) deletions++;
          }
        }

        final path = newPath ?? oldPath ?? '';
        files.add(DiffFile(
          path: path,
          oldPath: oldPath!,
          newPath: newPath!,
          status: _inferStatus(oldPath!, newPath!),
          additions: additions,
          deletions: deletions,
          hunks: List.unmodifiable(hunks),
        ));

        oldPath = null;
        newPath = null;
        hunks.clear();
      }
    }

    for (final line in lines) {
      // New file header signals start of a new file diff
      if (line.startsWith('--- ')) {
        final match = _fileHeaderOld.firstMatch(line);
        if (match != null) {
          flushFile();
          oldPath = match.group(1);
          continue;
        }
        // Handle /dev/null for new files
        if (line == '--- /dev/null') {
          flushFile();
          oldPath = '/dev/null';
          continue;
        }
      }

      if (line.startsWith('+++ ')) {
        final match = _fileHeaderNew.firstMatch(line);
        if (match != null) {
          newPath = match.group(1);
          continue;
        }
        if (line == '+++ /dev/null') {
          newPath = '/dev/null';
          continue;
        }
      }

      // Hunk header
      final hunkMatch = _hunkHeader.firstMatch(line);
      if (hunkMatch != null) {
        flushHunk();
        final oldStart = int.parse(hunkMatch.group(1)!);
        final oldCount = int.tryParse(hunkMatch.group(2) ?? '1') ?? 1;
        final newStart = int.parse(hunkMatch.group(3)!);
        final newCount = int.tryParse(hunkMatch.group(4) ?? '1') ?? 1;
        final header = line;

        currentHunk = DiffHunk(
          header: header,
          oldStart: oldStart,
          oldLines: oldCount,
          newStart: newStart,
          newLines: newCount,
          lines: const [],
        );
        currentHunkLines = [];
        oldLineNum = oldStart;
        newLineNum = newStart;
        continue;
      }

      // Diff content lines
      if (currentHunkLines != null) {
        if (line.startsWith('+')) {
          currentHunkLines!.add(DiffLine(
            type: DiffLineType.added,
            content: line.substring(1),
            newLineNumber: newLineNum++,
          ));
        } else if (line.startsWith('-')) {
          currentHunkLines!.add(DiffLine(
            type: DiffLineType.removed,
            content: line.substring(1),
            oldLineNumber: oldLineNum++,
          ));
        } else if (line.startsWith(' ')) {
          currentHunkLines!.add(DiffLine(
            type: DiffLineType.context,
            content: line.substring(1),
            oldLineNumber: oldLineNum++,
            newLineNumber: newLineNum++,
          ));
        }
      }
    }

    flushFile();
    return files;
  }

  static FileChangeStatus _inferStatus(String oldPath, String newPath) {
    if (oldPath == '/dev/null') return FileChangeStatus.added;
    if (newPath == '/dev/null') return FileChangeStatus.deleted;
    if (oldPath != newPath) return FileChangeStatus.renamed;
    return FileChangeStatus.modified;
  }
}
